File: siden-core/src/main/java/ninja/siden/internal/RoutingHandler.java
/*
 * Copyright 2014 SATO taichi
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */
package ninja.siden.internal;

import io.undertow.Undertow;
import io.undertow.predicate.Predicate;
import io.undertow.server.HttpHandler;
import io.undertow.server.HttpServerExchange;
import ninja.siden.*;
import ninja.siden.def.ErrorCodeRoutingDef;
import ninja.siden.def.ExceptionalRoutingDef;
import org.xnio.OptionMap;

import java.util.*;

/**
 * @author taichi
 */
public class RoutingHandler implements HttpHandler {

    static final Set<String> ignorePackages = new HashSet<>();

    static {
        ignorePackages.add(App.class.getPackage().getName());
        ignorePackages.add(Undertow.class.getPackage().getName());
    }

    final List<Routing> routings = new ArrayList<>();

    final Map<Class<? extends Throwable>, ExceptionalRoutingDef<? extends Throwable>> exceptionalMappings = new HashMap<>();

    final Map<Integer, List<ErrorCodeRoutingDef>> errorCodeMappings = new HashMap<>();

    final HttpHandler next;

    public RoutingHandler(HttpHandler handler) {
        this.next = handler;
    }

    @Override
    public void handleRequest(HttpServerExchange exchange) throws Exception {
        for (Routing route : routings) {
            if (route.predicate.resolve(exchange)) {
                HttpHandler hh = ex -> handle(ex, route.route::handle,
                        route.renderer);
                if (exchange.isInIoThread()) {
                    exchange.dispatch(hh);
                } else {
                    hh.handleRequest(exchange);
                }
                return;
            }
        }
        this.next.handleRequest(exchange);
    }

    boolean handle(Integer responseCode, HttpServerExchange exchange)
            throws Exception {
        if (exchange.isRequestChannelAvailable()) {
            exchange.setResponseCode(responseCode);
            List<ErrorCodeRoutingDef> list = this.errorCodeMappings
                    .getOrDefault(responseCode, Collections.emptyList());
            for (ErrorCodeRoutingDef route : list) {
                if (handle(exchange, route.getRoute()::handle, route.getRenderer())) {
                    return true;
                }
            }
        }
        return false;
    }

    <T extends Throwable> ExceptionalRoutingDef<T> find(T exception) {
        for (Class<?> c = exception.getClass(); c != null; c = c
                .getSuperclass()) {
            @SuppressWarnings("unchecked")
            ExceptionalRoutingDef<T> route = (ExceptionalRoutingDef<T>) this.exceptionalMappings.get(c);
            if (route != null) {
                return route;
            }
        }
        return null;
    }

    <T extends Throwable> boolean handle(T exception,
                                         HttpServerExchange exchange) throws Exception {
        ExceptionalRoutingDef<T> def = find(exception);
        return def != null && handle(exchange, (req, res) -> {
            exchange.setResponseCode(500);
            return def.getRoute().handle(exception, req, res);
        }, def.getRenderer());
    }

    boolean handle(HttpServerExchange exchange, Route fn, Renderer<?> renderer)
            throws Exception {
        try {
            Request request = exchange.getAttachment(Core.REQUEST);
            Response response = exchange.getAttachment(Core.RESPONSE);
            Object model = fn.handle(request, response);
            if (model != null) {
                if (model instanceof Optional) {
                    @SuppressWarnings("unchecked")
                    Optional<Object> opt = (Optional<Object>) model;
                    model = opt.map(v -> v).orElse(null);
                }
                if (model instanceof Integer) {
                    return handle((Integer) model, exchange);
                }
                if (model instanceof ExchangeState) {
                    return true;
                }
                if (model != null && contains(model.getClass()) == false) {
                    resolve(renderer, exchange).render(model, exchange);
                    return true;
                }
            }
            return handle(exchange.getResponseCode(), exchange);
        } catch (Exception ex) {
            if (exchange.isRequestChannelAvailable()) {
                if (handle(ex, exchange)) {
                    return true;
                }
            }
            throw ex;
        }
    }

    @SuppressWarnings("unchecked")
    Renderer<Object> resolve(Renderer<?> renderer, HttpServerExchange exchange) {
        if (renderer == null) {
            OptionMap config = exchange.getAttachment(Core.CONFIG);
            return config.get(Config.DEFAULT_RENDERER);
        }
        return (Renderer<Object>) renderer;
    }

    boolean contains(Class<?> clazz) {
        String n = clazz.getName();
        return ignorePackages.stream().anyMatch(n::startsWith);
    }

    public void add(Predicate predicate, Route route, Renderer<?> renderer) {
        this.routings.add(new Routing(predicate, route, renderer));
    }

    public void add(ExceptionalRoutingDef<?> model) {
        this.exceptionalMappings.put(model.getType(), model);
    }

    public void add(ErrorCodeRoutingDef model) {
        List<ErrorCodeRoutingDef> list = this.errorCodeMappings.get(model
                .getCode());
        if (list == null) {
            list = new ArrayList<>();
        }
        list.add(model);
        this.errorCodeMappings.put(model.getCode(), list);
    }

    class Routing {
        Predicate predicate;
        Route route;
        Renderer<?> renderer;

        Routing(Predicate predicate, Route route, Renderer<?> renderer) {
            this.predicate = predicate;
            this.route = route;
            this.renderer = renderer;
        }
    }
}

File: siden-core/src/main/kotlin/ninja/siden/internal/RoutingHandler.kt
/*
 * Copyright 2014 SATO taichi
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */
package ninja.siden.internal

import io.undertow.Undertow
import io.undertow.predicate.Predicate
import io.undertow.server.HttpHandler
import io.undertow.server.HttpServerExchange
import ninja.siden.App
import ninja.siden.Config
import ninja.siden.Renderer
import ninja.siden.Route
import ninja.siden.def.ErrorCodeRoutingDef
import ninja.siden.def.ExceptionalRoutingDef
import java.util.*
import kotlin.reflect.KClass

/**
 * @author taichi
 */
class RoutingHandler(internal val next: HttpHandler) : HttpHandler {

    internal val routings: MutableList<Routing> = ArrayList()

    internal val exceptionalMappings: MutableMap<Class<out Throwable>, ExceptionalRoutingDef<out Throwable>> = HashMap()

    internal val errorCodeMappings = HashMap<Int, MutableList<ErrorCodeRoutingDef>>()

    override fun handleRequest(exchange: HttpServerExchange) {
        for (route in routings) {
            if (route.predicate.resolve(exchange)) {
                val hh = HttpHandler { ex ->
                    handle(ex, Route { request, response -> route.route.handle(request, response) },
                            route.renderer)
                }
                if (exchange.isInIoThread) {
                    exchange.dispatch(hh)
                } else {
                    hh.handleRequest(exchange)
                }
                return
            }
        }
        this.next.handleRequest(exchange)
    }

    internal fun handle(responseCode: Int, exchange: HttpServerExchange): Boolean {
        if (exchange.isRequestChannelAvailable) {
            exchange.responseCode = responseCode
            val list = this.errorCodeMappings[responseCode] ?:emptyList<ErrorCodeRoutingDef>()
            for (route in list) {
                if (handle(exchange, Route { request, response -> route.route.handle(request, response) }, route.renderer)) {
                    return true
                }
            }
        }
        return false
    }

    internal fun <T : Throwable> find(exception: T): ExceptionalRoutingDef<T>? {
        var c: Class<*>? = exception.javaClass
        while (c != null) {
            @Suppress("UNCHECKED_CAST")
            val route = this.exceptionalMappings[c] as ExceptionalRoutingDef<T>?
            if (route != null) {
                return route
            }
            c = c.superclass
        }
        return null
    }

    internal fun <T : Throwable> handle(exception: T,
                                        exchange: HttpServerExchange): Boolean {
        val def = find(exception)
        return def != null && handle(exchange, Route { req, res ->
            exchange.responseCode = 500
            def.route.handle(exception, req, res)
        }, def.renderer)
    }

    internal fun handle(exchange: HttpServerExchange, fn: Route, renderer: Renderer<*>?): Boolean {
        try {
            val request = exchange.getAttachment(Core.REQUEST)
            val response = exchange.getAttachment(Core.RESPONSE)
            var model: Any? = fn.handle(request, response)
            if (model != null) {
                if (model is Optional<*>) {
                    model = model.orElse(null)
                }
                if (model is Int) {
                    return handle(model, exchange)
                }
                if (model is ExchangeState) {
                    return true
                }
                if (model != null && contains(model.javaClass) == false) {
                    resolve(renderer, exchange).render(model, exchange)
                    return true
                }
            }
            return handle(exchange.responseCode, exchange)
        } catch (ex: Exception) {
            if (exchange.isRequestChannelAvailable) {
                if (handle(ex, exchange)) {
                    return true
                }
            }
            throw ex
        }

    }

    @Suppress("UNCHECKED_CAST")
    internal fun resolve(renderer: Renderer<*>?, exchange: HttpServerExchange): Renderer<Any> {
        if (renderer == null) {
            val config = exchange.getAttachment(Core.CONFIG)
            return config.get(Config.DEFAULT_RENDERER) as Renderer<Any>
        }
        return renderer as Renderer<Any>
    }

    internal operator fun contains(clazz: Class<*>): Boolean {
        val n = clazz.name
        return ignorePackages.any( { n.startsWith(it) })
    }

    fun add(predicate: Predicate, route: Route, renderer: Renderer<*>?) {
        this.routings.add(Routing(predicate, route, renderer))
    }

    fun add(model: ExceptionalRoutingDef<*>) {
        this.exceptionalMappings.put(model.type, model)
    }

    fun add(model: ErrorCodeRoutingDef) {
        var list = this.errorCodeMappings[model.code]
        if (list == null) {
            list = ArrayList<ErrorCodeRoutingDef>()
        }
        list.add(model)
        this.errorCodeMappings.put(model.code, list)
    }

    internal data class Routing(var predicate: Predicate, var route: Route, var renderer: Renderer<*>?)

    companion object {

        internal val ignorePackages: MutableSet<String> = HashSet()

        internal val KClass<*>.packageName: String get() = this.java.`package`.name

        init {
            ignorePackages.add(App::class.packageName)
            ignorePackages.add(Undertow::class.packageName)
        }
    }
}
